### ISSUE TITLE
When I click on select2 my page scrolls to top

### ISSUE THREAD

**Issue by skazhikadyadya** [select2/select2#5022]
Hello. When I click on select2 dropdown on my page page scroll to top. I was tested in Mozilla Firefox and latest Google Chrome. Hown I can solve that problem?

**Comment by etharshrouf** [comment]
This problem happened with me too

**Comment by alexweissman** [comment]
Could you please check the open PRs to see if a patch has already been submitted?  If not, could one of you please create a jsfiddle to demonstrate this behavior?

**Comment by lk77** [comment]
the entire page or just the dropdown ?

**Comment by etharshrouf** [comment]
when click in selectbox first time the page scroll to top, but when scroll down to selectbox and click on selectbox page not scroll, but when make refresh and click to select the page scroll again

**Comment by alexweissman** [comment]
Please create a fully-functioning demonstration of this behavior (jsfiddle is fine).

**Comment by etharshrouf** [comment]
https://jsfiddle.net/fL9fj2sL/12/

**Comment by alexweissman** [comment]
Could you try adding some linebreaks or something to this fiddle?  As it is, this page doesn't seem to scroll at all.

**Comment by etharshrouf** [comment]
http://ts.com.ps/toof/toof/buyer-individual.html

**Comment by etharshrouf** [comment]
http://ts.com.ps/toof/toof/buyer-institute.html

**Comment by alexweissman** [comment]
Thanks. This appears to be a duplicate of #4810.  I've tested your page and can confirm this behavior on your page.

It's hard to say what's going on here - I don't see any event handlers other than select2 and jQuery attached to that element, so I don't think it would be interference from another plugin.  I do notice that you're using jQuery 3, as is the case in #4810.  Could you try (temporarily) downgrading to jQuery 2.x, and see if the problem persists?

**Comment by etharshrouf** [comment]
i downgrading jquery version to 2.x , and problem not solved

**Comment by OlegKi** [comment]
I wanted just make the demo https://jsfiddle.net/fL9fj2sL/12/ created by **etharshrouf** (see [the comment](https://github.com/select2/select2/issues/5022#issuecomment-326696661))  more clear.

The modified demo https://jsfiddle.net/OlegKi/fL9fj2sL/15/ uses jQuery 1.12.4, to close any additional questions about the version of jQuery, and it initializes select2 only after clicking on the button "Init select2". One can see that the source `<select>`  element have the position not on top of the page. I used `margin-top: 2000px;` to change the position of the `<select>` , but the same problem exist by usage other CSS properties (like `position: relative; top: 2000px;` and so on). One can click on the button "Init select2" and see that the select2 controls will be created on the wrong position directly under the button "Init select2".

I think that one can fix the problem in about the following way: One has to get the position of the source `<select>`  element with respect of `getBoundingClientRect()` or `jQuery.offset()` before initializing of select2 and to use the same position for select2. See https://jsfiddle.net/OlegKi/fL9fj2sL/16/. Of cause one should include the corresponding code fragments as the part of the code of select2.

**Comment by alexweissman** [comment]
@etharshrouf thanks for checking this for me.

@OlegKi is this the same issue, though?  The issue you're demonstrating has to do with the position of the control itself, not the window scroll position.

**Comment by OlegKi** [comment]
@alexweissman Yes, Alex. I think also that the reason of the problem is the wrong calculated position of the  `span.select2` and not the window scroll position. At least I didn't see any problem with the scroll position.

**Comment by alexweissman** [comment]
Thanks.  Could you check the current open and closed PRs and see if there are any which address this problem?

**Comment by alexweissman** [comment]
@maximevaillancourt @BillFilterEasy are you still experiencing this issue?

**Comment by OlegKi** [comment]
@alexweissman: I din't found an opened PRs about the subject. I used the latest GitHub sources (`https://rawgit.com/select2/select2/master/dist/js/select2.js` and  `https://rawgit.com/select2/select2/master/dist/css/select2.css`) in the demos https://jsfiddle.net/OlegKi/fL9fj2sL/15/ and https://jsfiddle.net/OlegKi/fL9fj2sL/16/ to be sure that the bug still exist in the latest version of select2. Thus the but really still exist and the usage of `getBoundingClientRect()` or/and `jQuery.offset()` allows to fix it.

**Comment by alexweissman** [comment]
Ahh, ok I am starting to understand the issue.  Question: does setting the [`dropdownParent`](http://select2.github.io/options.html#dropdownParent) fix this problem?

**Comment by alexweissman** [comment]
Here, try this example:

https://jsfiddle.net/nm4o7h10/1/

**Comment by OlegKi** [comment]
@alexweissman: I think that the usage of `dropdownParent` can't solve the problem. Your last demo https://jsfiddle.net/nm4o7h10/1/ works *not* because of the usage of `dropdownParent` option, but because of CSS setting on *another* element: on `<div class="select-container">` instead of on `<select id="js-example-basic-single">` itself.

You can just comment `dropdownParent: $('.select-container')`  in your last demo and to get https://jsfiddle.net/nm4o7h10/2/, which works too (has no bug). On the other side you can replace
```css
.select-container {
    position: relative;
    top: 2000px;
}
```
to
```css
#js-example-basic-single {
    position: relative;
    top: 2000px;
}
```
and we receive the original bug back even in case of usage `dropdownParent: $('.select-container')` option:
https://jsfiddle.net/nm4o7h10/3/

**Comment by alexweissman** [comment]
So perhaps the best solution to this problem is to simply avoid directly positioning the `<select>` control, and instead applying positions to a wrapper element instead?  Seems like a more elegant solution than messing around with `.offset`.

**Comment by OlegKi** [comment]
@alexweissman: Alex, you should finally decide, which features will be implemented in select2 and whether the developer, who uses select2, should solve some problems or if select2 will do that.

In my experience, select2 will be frequently used in *existing solution* to have more comfort for the users. One have already some solution, which displays `<select>` elements. One combines different plugins in the solution. For example, the `<select>` elements are inside of the grid/table, which is inside of jQuery UI dialog or Bootstrap dialog and so on. All CSS of all used plugins will be applied on the source `<select>` element. It's frequently that `position`, `margin-top`, `top` and many other CSS properties will be set *indirectly* during including of other CSS (Bootstrap CSS, jQuery UI CSS and so on). The `<select>` elements itself could be created *indirectly* by other plugins. Creating of wrappers for every `<select>` element could be not so easy in real world.

It would be much more comfortable (for developers, who usage select2) that select2 will be just created on *the same position*, where `<select>` element exists. The final decision is your choice. If you think about some possible side effects of usage `.offset`, then you can do this not by default and introduce a new option, which do this.

**Comment by alpesh88ww** [comment]
i found somthing in css if u have add like bellow please check :
if you puted this kindof line in you css please remove it.  **transition** in css on SPAN tag.
`span { -webkit-transition: all .2s ease-in-out; -moz-transition: all .2s ease-in-out; -o-transition: all .2s ease-in-out; transition: all .2s ease-in-out; }`

**Comment by Bleech94** [comment]
Just discovered this issue on my page where users fill out a long form with 20 or so select2 dropdowns. I Haven't been able to fix it myself using the recommendations above. The issue seems to happen pretty consistently when you click a dropdown, scroll, then click another one. Any progress on this?

**Comment by alexweissman** [comment]
Are there any open pull requests that address the problem?  If you find one, I can take a look.

**Comment by Bleech94** [comment]
I couldn't find any related. I can try and fix it and make a PR if you give me some idea of what you think the issue could be and how to fix it.

**Comment by alexweissman** [comment]
There is one idea involving `offset` and `getBoundingClientRect`:

> I think that one can fix the problem in about the following way: One has to get the position of the source <select> element with respect of getBoundingClientRect() or jQuery.offset() before initializing of select2 and to use the same position for select2. See https://jsfiddle.net/OlegKi/fL9fj2sL/16/. Of cause one should include the corresponding code fragments as the part of the code of select2.

I'm afraid I don't really have much insight on this issue otherwise, sorry.

**Comment by OlegKi** [comment]
@alexweissman I'll comment the demos https://jsfiddle.net/OlegKi/fL9fj2sL/15/ and https://jsfiddle.net/OlegKi/fL9fj2sL/16/ posted in my comment before:

The demo https://jsfiddle.net/OlegKi/fL9fj2sL/15/ shows the existing problem in select2. It has `<select class="js-example-basic-single">`, which `margin-top` is `2000px`. One have to scroll to the bottom to see the select. After clicking on the "Init select2" button the code `$(".js-example-basic-single").select2();` will be executed and the select2 will be created on the *wrong position* (directly under `<button>` instead on the same place where `<select>` exist).

Another demo https://jsfiddle.net/OlegKi/fL9fj2sL/16/ uses `var offset = $src.offset();` (or `$src[0].getBoundingClientRect()` instead of `$src.offset()`) to get the position of the `<select>`, which will be converted later to select2. Then the demo calls `.select2()` and finally it **changes the position of `$(".select2")`** so that the position of select2 will be the same as the position of original `<select>` .

**Comment by Bleech94** [comment]
Thanks @alexweissman 

I read all of the comments above, I'm not convinced that is the same issue @OlegKi. My select2s are being placed correctly, the problem is when I click a select2, scroll, then click another select2. Sometimes the page jumps in the direction of the last one selected. I don't think fixing your issue will fix mine because even if the select2s currently have the wrong position, fixing their position will still cause my page to jump towards the last select2.

If you can make a jsfiddle that shows my issue and how your able to fix it with your solution, I'll look into it more.

**Comment by sftl** [comment]
Hi everybody,

First of all, sorry for my bad english, but i will try to explain it the best i can.

This issue also happen to me just in firefox (Chrome and IE11 works perfect).
After 2 days, reviewing the source code....i found a solution that works for me.
In my case, it seems like the problem comes because of the default position of the dropdown before to be displayed to the user.

So the solution i found is (i'm in version 4.0.4) : 
 * In  AttachBody.prototype.position, comment that line "top: -999999" (line 4200 apox)

Hope this can help somebody.
Cheers.

**Comment by alexweissman** [comment]
Please avoid posting "hacks" to the core codebase.  If you have a thoroughly tested solution, you should submit a pull request.  As it stands, you haven't provided any context for this change, and there is no way to know what unintended side effects it might have.

**Comment by lk77** [comment]
Hello,

i've checked the call tree of select2 to find where this problem is triggered and i found a call to this piece of code (line 877 on select2.full.js v 4.0.5) :
```
Results.prototype.highlightFirstItem = function () {
                var $options = this.$results
                    .find('.select2-results__option[aria-selected]');

                var $selected = $options.filter('[aria-selected=true]');

                // Check if there are any selected options
                if ($selected.length > 0) {
                    // If there are selected options, highlight the first
                    $selected.first().trigger('mouseenter');
                } else {
                    // If there are no selected options, highlight the first option
                    // in the dropdown
                    $options.first().trigger('mouseenter');
                }

                this.ensureHighlightVisible();
            };
```
this seems to be the source of the problem, especially the  $selected.first().trigger('mouseenter') part
if i comment this line, the problem is gone.

**Comment by pingcheng** [comment]
Hi, I have this issues too, but I just found it conflicts with the CSS transition, I applied the transition animation to * selector, and that property breaks select2.

I removed the transition and everything works fine again.

**Comment by ko22009** [comment]
How I can fix it?

**Comment by rpujakesuma** [comment]
Hi I have the solutions.. add this on your css 
`* {
	-webkit-transition: none !important;
	   -moz-transition: none !important;
	    -ms-transition: none !important;
	     -o-transition: none !important;
	        transition: none !important;
}`

I hope this help everyone on the future..

**Comment by cheklame** [comment]
Hii every one,
I' m using selcet2 version 4.0.6 and this proplem persists, and with your solution i can't fix my problem,
I'm beginner in the javaScript and visual studio, i m using asp.net core i have the same problem i tried to edit the select2.js but it seams like it is not compiling, i can't see any changes, my question is how can i edit the select2.js and how can i compile it to see the changes.
thak you for your help.

**Comment by cheklame** [comment]
Hi,
Can you tell how how to edit and compile select2.js file in asp.net core project thank you

**Comment by mprandot** [comment]
same problem here. @rpujakesuma css not work for me

**Comment by tuanbl** [comment]
@rpujakesuma exellent

**Comment by AndreiShenets** [comment]
I have this issue in IE 11 on select 2 4.0.4. version and I fixed like following in S2.define('select2/dropdown/attachBody'. code:


```line:4217
                this.$dropdownContainer.on('mousedown', function (evt) {
                    console.log('evt.stopPropagation();')
                    evt.stopPropagation();
                    return false;
                });
```

**Comment by felipe-look** [comment]
I'm still having this problem, but not in all pages with select2, and I can't identify the problem yet.

**Comment by husnainahmed61** [comment]
same issue,
css not working for me

**Comment by felipe-look** [comment]
I solving by removing the span CSS transition
![image](https://user-images.githubusercontent.com/13525019/65530981-3d689d80-def9-11e9-8cde-8b6c9345852d.png)

**Comment by aminkariminasab** [comment]
find $dropdownContainer.remove() in select2.js and then change remove top in the manual style.

**Comment by aliarslan93** [comment]
> Hi I have the solutions.. add this on your css
> 
> > .select2-container, .select2-dropdown, .select2-search, .select2-results {
> > -webkit-transition: none !important;
> > -moz-transition: none !important;
> > -ms-transition: none !important;
> > -o-transition: none !important;
> > transition: none !important;
> > }
> 
> I hope this help everyone on the future..

Thanks <3

**Comment by carlosromanxyz** [comment]
> Hi I have the solutions.. add this on your css
> 
> > .select2-container, .select2-dropdown, .select2-search, .select2-results {
> > -webkit-transition: none !important;
> > -moz-transition: none !important;
> > -ms-transition: none !important;
> > -o-transition: none !important;
> > transition: none !important;
> > }
> 
> I hope this help everyone on the future..

This solved my day <3

**Comment by vinodjosephaus** [comment]
Ensure the select control has an id attribute specified.  That worked for me.

In my case:
There were multiple select controls in my form and the initial code was:
<select name="countries[]" class="select2"> - WON'T work

Changed it to:
<select id="country' . $i . '" name="countries[]" class="select2"> - WORKS
(where $i is a variable which will ensure the ids are unique)

**Comment by BFalbo** [comment]
Hey guys, I had a similar problem. 

In my case, when I clicked out of the select2 or selected an option (so when the dropdown closed), my page scrolled to the bottom.

Tried everything, what solved was remove the search box from the select2:

$('.select2').select2({minimumResultsForSearch: -1});

Hope that can help you some way

**Comment by SupunSpera** [comment]
> Hi I have the solutions.. add this on your css
> 
> > .select2-container, .select2-dropdown, .select2-search, .select2-results {
> > -webkit-transition: none !important;
> > -moz-transition: none !important;
> > -ms-transition: none !important;
> > -o-transition: none !important;
> > transition: none !important;
> > }
> 
> I hope this help everyone on the future..

Fixed from this TY

**Comment by guptanilaj** [comment]
same problem happened with me.

**Comment by nnujones** [comment]
> Hi I have the solutions.. add this on your css
> 
> > .select2-container, .select2-dropdown, .select2-search, .select2-results {
> > -webkit-transition: none !important;
> > -moz-transition: none !important;
> > -ms-transition: none !important;
> > -o-transition: none !important;
> > transition: none !important;
> > }
> 
> I hope this help everyone on the future..

Still a working solution as of April 2026. Thank you!
